src/osd/ECBackend.cc

```diff
@@ -1272,37 +1272,19 @@ void ECBackend::filter_read_op(
 {
   set<hobject_t> to_cancel;
   for (map<pg_shard_t, set<hobject_t> >::iterator i = op.source_to_obj.begin();
-       i != op.source_to_obj.end();
-       ++i) {
+       i != op.source_to_obj.end();) {
     if (osdmap->is_down(i->first.osd)) {
       to_cancel.insert(i->second.begin(), i->second.end());
       op.in_progress.erase(i->first);
-      continue;
-    }
-  }
-
-  if (to_cancel.empty())
-    return;
-
-  for (map<pg_shard_t, set<hobject_t> >::iterator i = op.source_to_obj.begin();
-       i != op.source_to_obj.end();
-       ) {
-    for (set<hobject_t>::iterator j = i->second.begin();
-	 j != i->second.end();
-	 ) {
-      if (to_cancel.count(*j))
-	i->second.erase(j++);
-      else
-	++j;
-    }
-    if (i->second.empty()) {
       op.source_to_obj.erase(i++);
     } else {
-      assert(!osdmap->is_down(i->first.osd));
       ++i;
     }
   }
 
+  if (to_cancel.empty())
```

Comment:
This patch looks to me like it changes the behavior... the set<hobject_t>'s aren't getting filtered.

Reply:
@liewegas If a specific shard is down, we are going to erase that shard from the `source_to_obj` map anyway. I don't think we need to vacate the object set first...